infrastructure/templates/public-api/components/functionApp.bicep

```diff
@@ -42,51 +42,216 @@ param subnetId string
+param functionAppExists bool
+
 var appServicePlanName = '${resourcePrefix}-asp-${functionAppName}'
-var reserved = appServicePlanOS == 'Linux' ? true : false
-var functionName = '${resourcePrefix}-fa-${functionAppName}'
+var reserved = appServicePlanOS == 'Linux'
+var fullFunctionAppName = '${subscription}-ees-papi-fa-${functionAppName}'
 
 resource appServicePlan 'Microsoft.Web/serverfarms@2022-09-01' = {
   name: appServicePlanName
   location: location
-  kind: kind
+  kind: 'functionapp'
   sku: sku
   properties: {
     reserved: reserved
   }
 }
 
+var dedicatedStorageAccountName = replace('${subscription}eessa${functionAppName}', '-', '')
+
+// Create a dedicated Storage Account for this Function App for it to store its deployment code and jobs in.
+// Grant the Function App access by whitelisting its subnet for inbound traffic.
+//
+// For performance, it is considered good practice for each Function App to have its own dedicated Storage Account. See
+// https://learn.microsoft.com/en-us/azure/azure-functions/storage-considerations?tabs=azure-cli#optimize-storage-performance.
+resource dedicatedStorageAccount 'Microsoft.Storage/storageAccounts@2022-05-01' = {
+  name: dedicatedStorageAccountName
+  location: location
+  sku: {
+    name: 'Standard_LRS'
+  }
+  kind: 'Storage'
+  properties: {
+    supportsHttpsTrafficOnly: true
+    defaultToOAuthAuthentication: true
+    networkAcls: {
+      bypass: 'AzureServices'
+      defaultAction: 'Deny'
+      virtualNetworkRules: [
+        {
+          action: 'Allow'
+          id: subnetId
+        }
+      ]
+    }
+  }
+}
+
+var dedicatedStorageAccountString = 'DefaultEndpointsProtocol=https;AccountName=${dedicatedStorageAccountName};EndpointSuffix=${environment().suffixes.storage};AccountKey=${dedicatedStorageAccount.listKeys().keys[0].value}'
+
+// When deploying to a Function App utilising a secured VNet to store its deployment files,
+// unique file shares must be pre-generated and unique to each slot prior to deploying the
+// Function App itself if we wish to use slot swapping.
+//
+// See the second paragraph of https://learn.microsoft.com/en-us/azure/azure-functions/functions-infrastructure-as-code?tabs=json%2Clinux%2Cdevops&pivots=premium-plan#secured-deployments.
+var fileShareName1 = '${toLower(fullFunctionAppName)}-1'
+var fileShareName2 = '${toLower(fullFunctionAppName)}-2'
+
+resource fileShare1 'Microsoft.Storage/storageAccounts/fileServices/shares@2022-05-01' = {
+  name: '${dedicatedStorageAccountName}/default/${fileShareName1}'
+  dependsOn: [
+    dedicatedStorageAccount
+  ]
+}
+
+resource fileShare2 'Microsoft.Storage/storageAccounts/fileServices/shares@2022-05-01' = {
+  name: '${dedicatedStorageAccountName}/default/${fileShareName2}'
+  dependsOn: [
+    dedicatedStorageAccount
+  ]
+}
+
 resource functionApp 'Microsoft.Web/sites@2023-01-01' = {
```

Comment:
We could run this as a containerised Function app within the Container App environment which might be more consistent because then all our apps are containerised without slots, as apposed to a mixture of apps, i.e. container apps without slots and app services running code deploys with slots.

I'm just wondering if this was already considered?

Reply:
:grin: :grimacing: